[#3780] Handle ChannelInitializer exception in exceptionCaught()

normanmaurer opened the pull request:
Motivation:

At the moment we directly closed the Channel when an exception accoured durring initChannel(...) without giving the user any way to do extra or special handling.

Modifications:

Handle the exception in exceptionCaught(...) of the ChannelInitializer which will by default log and close the Channel. This way the user can override this.

Result:

More felixible handling of exceptions.

normanmaurer commented:
@trustin please check

Scottmitch commented on transport/src/test/java/io/netty/channel/DefaultChannelPipelineTest.java:
```diff
@@ -535,6 +536,28 @@ public void testLastHandlerEmptyPipeline() throws Exception {
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelInitializer<Channel>() {
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                throw new IllegalStateException();
```
Instantiate this exception outside the channel scope so you can do `assertSame` at the end of the test?

trustin commented:
LGTM. Please address @Scottmitch's comment before cherry-picking though. :-)

normanmaurer commented:
Addressed comment and cherry-picked into 4.0 (c9e71aafbd58b3fed347ba5cd1d931074264f64e) , 4.1 (0f6559d188f46be591902091f8150ad36c655c4f) and master (8d25aab00e383ed586140d204045a8e2854ded08)
